### ISSUE TITLE
ServerModule + NoopAnimationsModule - Cannot read property 'display' of undefined at cloakElement

### ISSUE THREAD

**Issue by dmitrysteblyuk** [angular/angular#18562]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
When prerendering application with animated components this error happens despite that NoopAnimationsModule is provided on server side instead of BrowserAnimationsModule. I use @angular/material but looking at this stack trace I believe it's hardly relevant here:

```
ERROR TypeError: Cannot read property 'display' of undefined
    at cloakElement (\node_modules\@angular\animations\bundles\animations-browser.umd.js:4374:50)
    at \node_modules\@angular\animations\bundles\animations-browser.umd.js:4386:79
    at Array.map (native)
    at cloakAndComputeStyles (\node_modules\@angular\animations\bundles\animations-browser.umd.js:4386:47)
    at TransitionAnimationEngine._flushAnimations (\node_modules\@angular\animations\bundles\animations-browser.umd.js:3880:46)
    at TransitionAnimationEngine.flush (\node_modules\@angular\animations\bundles\animations-browser.umd.js:3703:32)
    at InjectableAnimationEngine.AnimationEngine.flush (\node_modules\@angular\animations\bundles\animations-browser.umd.js:4642:32)
    at \node_modules\@angular\platform-browser\bundles\platform-browser-animations.umd.js:333:30
    at ZoneDelegate.invoke (\node_modules\zone.js\dist\zone-node.js:391:26)
    at Zone.run (\node_modules\zone.js\dist\zone-node.js:141:43)
    at NgZone.runOutsideAngular (\node_modules\@angular\core\bundles\core.umd.js:3876:83)
    at AnimationRendererFactory.end (\node_modules\@angular\platform-browser\bundles\platform-browser-animations.umd.js:331:24)
    at ViewRef_.detectChanges (\node_modules\@angular\core\bundles\core.umd.js:10203:16)
    at \node_modules\@angular\core\bundles\core.umd.js:4839:63
    at Array.forEach (native)
    at ApplicationRef_.tick (\node_modules\@angular\core\bundles\core.umd.js:4839:25)
```

## Expected behavior
Those functions should not be called on server side since ```style``` is not defined on Element there.

## Minimal reproduction of the problem with instructions
I believe it can be reproduced by calling platformServer.renderModuleFactory(AppModule) if AppModule contains ServerModule, NoopAnimationsModule and some component with animations

## What is the motivation / use case for changing the behavior?
Would be nice to avoid polluting logs with these errors

## Environment

<pre><code>
Angular version: 4.3.3

Browser: any
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.0.0  
- Platform:  Win 10 

Others:

</code></pre>

**Comment by kbarralon** [comment]
Same here

**Comment by Kamshak** [comment]
Getting the same thing, even without an animated component. This causes it for me:
```ts
<div class="router-loading-indicator" *ngIf="routerLoading$ | async">
    <spinner></spinner>
</div>
```
inside of my app component template.

**Comment by funkizer** [comment]
Same error. I'm not using animations or even importing any animation module. Am using Material though (button, dialog, input, select, card,...).

**Comment by funkizer** [comment]
Hack: Edited animations-browser.umd.js in server node_modules/@angular/animations/bundles
function cloakElement(element, value) {
        if (!element.style) { // ADDED
            element.style = { display: 'block' };
        }
        var /** @type {?} */ oldValue = element.style.display;
        element.style.display = value != null ? value : 'none';
        return oldValue;
    }

This got rid of the undefined display error, but now I'm getting below error on Firebase Cloud functions. Everything runs fine stand-alone in Node or with "firebase serve --only functions". 

TypeError: ctor is not a constructor
    at _createClass (/user_code/node_modules/@angular/core/bundles/core.umd.js:9544:26)
    at _createProviderInstance$1 (/user_code/node_modules/@angular/core/bundles/core.umd.js:9519:26)
    at initNgModule (/user_code/node_modules/@angular/core/bundles/core.umd.js:9473:13)
    at new NgModuleRef_ (/user_code/node_modules/@angular/core/bundles/core.umd.js:10587:9)
    at createNgModuleRef (/user_code/node_modules/@angular/core/bundles/core.umd.js:10571:12)
    at Object.debugCreateNgModuleRef [as createNgModuleRef] (/user_code/node_modules/@angular/core/bundles/core.umd.js:12865:12)
    at NgModuleFactory_.create (/user_code/node_modules/@angular/core/bundles/core.umd.js:13860:25)
    at /user_code/node_modules/@angular/core/bundles/core.umd.js:4529:61
    at ZoneDelegate.invoke (/user_code/node_modules/zone.js/dist/zone-node.js:391:26)
    at Object.onInvoke (/user_code/node_modules/@angular/core/bundles/core.umd.js:3922:33)

**Comment by josephliccini** [comment]
Same problem here.  I am importing NoopAnimationsModule into my AppServerModule and I thought that would take care of it, but it appears to still try and run some `style` modification on Node..

**Comment by AndreyZelenskiy** [comment]
@vikerman Do you have any updates or workaround for this issue?

**Comment by vikerman** [comment]
Switched to using Domino in 5.0 which should provide a more of the DOM API on the server. Can someone verify whether this is still an issue?

**Comment by chencingyen** [comment]
@funkizer thanks for your solution... it works for me :)

btw any ide to do it inside our project-file?

**Comment by kemmis** [comment]
Same problem here. I am using Angular Material, but the error doesn't appear to be related. I'm getting the error when conditionally hiding a div with *ngIf like this:
`<div class="pp-no-comments" *ngIf="comments.length == 0">
        No comments yet!
    </div>`

**Comment by kakins** [comment]
I'm getting the error when even declaring an import statement in a component:
import { trigger, style, transition, animate } from '@angular/animations';

I do not get that error if I remove the import statement.

I also have NoopAnimationsModule declared in my app server module.

**Comment by JayChase** [comment]
@vikerman this is still an issue. Is the solution from @funkizer going into an upcoming 4.x release? Also could you clarify whether the fact that the cloakElement function being called on the server means that somehow the browser animations module is still being imported into the server side bundle?

**Comment by naveedahmed1** [comment]
Any update on this?

**Comment by funkizer** [comment]
I haven't had this issue for a while and I've updated&reinstalled all nodemodules, so my hack has to be gone. Are you using angular2-universal-polyfills or similar? I removed it as it's not needed anymore (even though a lot of tutorials still use it), maybe that did it?

**Comment by vikerman** [comment]
I will have time to look into this next week and get it into the next 4.x patch release

**Comment by zolakt** [comment]
I'm still getting this in 4.4.6

It's enough just to define an animation trigger, regardless of the animation:
```
@Component({
    animations: [trigger('routerAnimation', [
    ])],
    ...
})
```

Breaks with BrowserAnimationsModule and  NoopAnimationsModule

**Comment by vikerman** [comment]
This was fixed in Angular >= 5.0.0 but never made it below that.
